File: packages/api/src/endpoints/tx.ts
import type { TxDetailParams, TxDetailResponse, TxListParams, TxListResponse } from "@/types/txTypes";

import { handleFetch } from "@/lib/handleFetch";
import type { DrepRegistrationsResponse } from "@/types/drepTypes";
import type { PoolRegistrationsResponse } from "@/types/poolTypes";
import type { StakeRegistrationsResponse } from "@/types/stakeTypes";
import type { ContractInteractionsResponse } from "@/types/contractTypes";

/**
 * Fetches detailed information about a specific transaction by its hash.
 *
 * @param {TxDetailParams} params - Parameters containing the transaction hash.
 * @param {string} params.hash - The transaction hash to fetch details for.
 *
 * @returns {Promise<TxDetailResponse>} A promise resolving to the transaction detail data.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getTxDetail = async ({ hash }: TxDetailParams) => {
  const url = `/tx/detail?hash=${hash}`;

  return handleFetch<TxDetailResponse>(url);
};

/**
 * Fetches a paginated list of transactions with optional filters.
 *
 * @param {TxListParams} params - Parameters for fetching transaction list.
 * @param {string} [params.hash] - Filter transactions by hash.
 * @param {number} [params.limit=10] - Maximum number of transactions to fetch.
 * @param {number} [params.offset=0] - Pagination offset.
 * @param {string} [params.address] - Filter transactions by address.
 * @param {string} [params.stake] - Filter transactions by stake.
 * @param {string} [params.asset] - Filter transactions by asset.
 * @param {string} [params.script] - Filter transactions by script.
 * @param {boolean} [params.has_donation] - Filter transactions by donation presence.
 * @param {string} [params.policy] - Filter transactions by policy.
 *
 * @returns {Promise<TxListResponse>} A promise resolving to the filtered list of transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getTxList = async ({ hash, limit = 10, offset = 0, address, stake, asset, script, has_donation, policy }: TxListParams) => {
  const url = `/tx/list`;
  const options = {
    params: {
      limit,
      offset,
      hash,
      address,
      stake,
      asset,
      script,
      has_donation,
      policy,
    },
  };

  return handleFetch<TxListResponse>(url, offset, options);
};

/**
 * Fetches a paginated list of DRep registration transactions.
 *
 * @param {TxListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of records to return.
 * @param {number} [params.offset=0] - Number of records to skip for pagination.
 *
 * @returns {Promise<DrepRegistrationsResponse>} A promise resolving to the list of DRep registration transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getDrepRegistrations = async ({ limit = 10, offset = 0 }: TxListParams) => {
  const url = `/tx/filter`;
  const options = {
    params: {
      limit,
      offset,
      type: "drep_registrations",
    },
  };

  return handleFetch<DrepRegistrationsResponse>(url, offset, options);
};

/**
 * Fetches a paginated list of DRep deregistration transactions.
 *
 * @param {TxListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of records to return.
 * @param {number} [params.offset=0] - Number of records to skip for pagination.
 *
 * @returns {Promise<DrepRegistrationsResponse>} A promise resolving to the list of DRep deregistration transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getDrepDeregistrations = async ({ limit = 10, offset = 0 }: TxListParams) => {
  const url = `/tx/filter`;
  const options = {
    params: {
      limit,
      offset,
      type: "drep_deregistrations",
    },
  };

  return handleFetch<DrepRegistrationsResponse>(url, offset, options);
};

/**
 * Fetches a paginated list of DRep update transactions.
 *
 * @param {TxListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of records to return.
 * @param {number} [params.offset=0] - Number of records to skip for pagination.
 *
 * @returns {Promise<DrepRegistrationsResponse>} A promise resolving to the list of DRep update transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getDrepUpdates = async ({ limit = 10, offset = 0 }: TxListParams) => {
  const url = `/tx/filter`;
  const options = {
    params: {
      limit,
      offset,
      type: "drep_updates",
    },
  };

  return handleFetch<DrepRegistrationsResponse>(url, offset, options);
};

/**
 * Fetches a paginated list of pool registration transactions.
 *
 * @param {TxListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of records to return.
 * @param {number} [params.offset=0] - Number of records to skip for pagination.
 *
 * @returns {Promise<PoolRegistrationsResponse>} A promise resolving to the list of pool registration transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getPoolRegistrations = async ({ limit = 10, offset = 0 }: TxListParams) => {
  const url = `/tx/filter`;
  const options = {
    params: {
      limit,
      offset,
      type: "pool_registrations",
    },
  };

  return handleFetch<PoolRegistrationsResponse>(url, offset, options);
};

/**
 * Fetches a paginated list of pool deregistration transactions.
 *
 * @param {TxListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of records to return.
 * @param {number} [params.offset=0] - Number of records to skip for pagination.
 *
 * @returns {Promise<PoolRegistrationsResponse>} A promise resolving to the list of pool deregistration transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getPoolDeregistrations = async ({ limit = 10, offset = 0 }: TxListParams) => {
  const url = `/tx/filter`;
  const options = {
    params: {
      limit,
      offset,
      type: "pool_deregistrations",
    },
  };

  return handleFetch<PoolRegistrationsResponse>(url, offset, options);
};

/**
 * Fetches a paginated list of stake key registration transactions.
 *
 * @param {TxListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of records to return.
 * @param {number} [params.offset=0] - Number of records to skip for pagination.
 *
 * @returns {Promise<StakeRegistrationsResponse>} A promise resolving to the list of stake key registration transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getStakeRegistrations = async ({ limit = 10, offset = 0 }: TxListParams) => {
  const url = `/tx/filter`;
  const options = {
    params: {
      limit,
      offset,
      type: "stake_key_registrations",
    },
  };

  return handleFetch<StakeRegistrationsResponse>(url, offset, options);
};

/**
 * Fetches a paginated list of contract interaction transactions.
 *
 * @param {TxListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of records to return.
 * @param {number} [params.offset=0] - Number of records to skip for pagination.
 *
 * @returns {Promise<ContractInteractionsResponse>} A promise resolving to the list of contract transactions.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getContractTransactions = async ({ limit = 10, offset = 0 }: TxListParams) => {
  const url = `/tx/filter`;
  const options = {
    params: {
      limit,
      offset,
      type: "contract_transactions",
    },
  };

  return handleFetch<ContractInteractionsResponse>(url, offset, options);
};

File: packages/cexplorer-api-rs/src/endpoints/tx.rs
use crate::client::{fetch, fetch_with_params};
use crate::error::CexplorerError;
use crate::types::tx_types::*;
use crate::types::drep_types::DrepRegistrationsResponse;
use crate::types::pool_types::PoolRegistrationsResponse;
use crate::types::stake_types::StakeRegistrationsResponse;
use crate::types::contract_types::ContractInteractionsResponse;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct TxListParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub address: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stake: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub script: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub has_donation: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub policy: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TxFilterParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<u64>,
    #[serde(rename = "type")]
    pub filter_type: String,
}

pub async fn get_tx_detail(hash: &str) -> Result<TxDetailResponse, CexplorerError> {
    let endpoint = format!("/tx/detail?hash={}", hash);
    fetch::<TxDetailResponse>(&endpoint).await
}

pub async fn get_tx_list(
    hash: Option<&str>,
    limit: Option<u64>,
    offset: Option<u64>,
    address: Option<&str>,
    stake: Option<&str>,
    asset: Option<&str>,
    script: Option<&str>,
    has_donation: Option<bool>,
    policy: Option<&str>,
) -> Result<TxListResponse, CexplorerError> {
    let endpoint = "/tx/list";
    let params = TxListParams {
        hash: hash.map(|s| s.to_string()),
        limit,
        offset,
        address: address.map(|s| s.to_string()),
        stake: stake.map(|s| s.to_string()),
        asset: asset.map(|s| s.to_string()),
        script: script.map(|s| s.to_string()),
        has_donation,
        policy: policy.map(|s| s.to_string()),
    };
    fetch_with_params::<TxListResponse, TxListParams>(endpoint, Some(&params)).await
}

pub async fn get_drep_registrations(
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<DrepRegistrationsResponse, CexplorerError> {
    let endpoint = "/tx/filter";
    let params = TxFilterParams {
        limit,
        offset,
        filter_type: "drep_registrations".to_string(),
    };
    fetch_with_params::<DrepRegistrationsResponse, TxFilterParams>(endpoint, Some(&params)).await
}

pub async fn get_drep_deregistrations(
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<DrepRegistrationsResponse, CexplorerError> {
    let endpoint = "/tx/filter";
    let params = TxFilterParams {
        limit,
        offset,
        filter_type: "drep_deregistrations".to_string(),
    };
    fetch_with_params::<DrepRegistrationsResponse, TxFilterParams>(endpoint, Some(&params)).await
}

pub async fn get_drep_updates(
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<DrepRegistrationsResponse, CexplorerError> {
    let endpoint = "/tx/filter";
    let params = TxFilterParams {
        limit,
        offset,
        filter_type: "drep_updates".to_string(),
    };
    fetch_with_params::<DrepRegistrationsResponse, TxFilterParams>(endpoint, Some(&params)).await
}

pub async fn get_pool_registrations(
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<PoolRegistrationsResponse, CexplorerError> {
    let endpoint = "/tx/filter";
    let params = TxFilterParams {
        limit,
        offset,
        filter_type: "pool_registrations".to_string(),
    };
    fetch_with_params::<PoolRegistrationsResponse, TxFilterParams>(endpoint, Some(&params)).await
}

pub async fn get_pool_deregistrations(
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<PoolRegistrationsResponse, CexplorerError> {
    let endpoint = "/tx/filter";
    let params = TxFilterParams {
        limit,
        offset,
        filter_type: "pool_deregistrations".to_string(),
    };
    fetch_with_params::<PoolRegistrationsResponse, TxFilterParams>(endpoint, Some(&params)).await
}

pub async fn get_stake_registrations(
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<StakeRegistrationsResponse, CexplorerError> {
    let endpoint = "/tx/filter";
    let params = TxFilterParams {
        limit,
        offset,
        filter_type: "stake_key_registrations".to_string(),
    };
    fetch_with_params::<StakeRegistrationsResponse, TxFilterParams>(endpoint, Some(&params)).await
}

pub async fn get_contract_transactions(
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<ContractInteractionsResponse, CexplorerError> {
    let endpoint = "/tx/filter";
    let params = TxFilterParams {
        limit,
        offset,
        filter_type: "contract_transactions".to_string(),
    };
    fetch_with_params::<ContractInteractionsResponse, TxFilterParams>(endpoint, Some(&params)).await
}
